### ISSUE TITLE
`torch.compile` SEGFAULTs on a ZeroTensor produced by a normal autograd path (gradient of `torch.sgn`)

### ISSUE THREAD

**Issue by tinywisdom** [pytorch/pytorch#197067]
### 🐛 Describe the bug

A `ZeroTensor` produced by a **normal autograd path** — the gradient of `torch.sgn` — makes a
`torch.compile`d function **segfault** when the tensor is used (indexed) inside it. Eager runs
fine and returns the correct value. This is memory-unsafe (SIGSEGV in the generated kernel),
not a graceful error.

## Minimal repro

```python
import torch

# A ZeroTensor arises naturally as the gradient of sgn (derivative of sgn is 0):
x = torch.randn(3, requires_grad=True)
z = torch.autograd.grad(torch.sgn(x).sum(), x)[0]
assert z._is_zerotensor()          # True -- an efficient ZeroTensor, from plain autograd

def fn(t):
    return t[-1] + 10

print(fn(z))                                  # eager:    tensor(10.)
torch.compile(fn, backend="inductor")(z)      # compiled: SEGFAULT (SIGSEGV, exit 139)
```

Deterministic across runs on **CPU** (the inductor CPU C++ kernel segfaults). On CUDA the
compiled path does not segfault (it completes), so this is a CPU-inductor issue.

## Expected behavior

Compiling a function that reads a ZeroTensor should behave like eager (a ZeroTensor reads as
zeros), or raise a graceful error — it must not segfault. A compiler should never turn a valid
eager program into a memory-unsafe crash.

## Actual behavior

Eager returns `tensor(10.)`. Under `torch.compile` (inductor), the generated kernel reads the
ZeroTensor's absent data pointer and the process crashes with SIGSEGV. So a value produced by
ordinary autograd (`grad` of `sgn`), passed into a compiled function, crashes the process.

### Versions

```
PyTorch version: 2.15.0a0+gite67e43b
Is debug build: True
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 12.3.0-1ubuntu1~22.04.3) 12.3.0
Clang version: 17.0.6
CMake version: version 4.0.0
Libc version: glibc-2.35

Python version: 3.12.13 | packaged by conda-forge (64-bit runtime)
Python platform: Linux-6.8.0-138-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.6.20
GPU 0: NVIDIA RTX A6000
Nvidia driver version: 580.82.07
cuDNN version: 9.3.0

[pip3] numpy==2.3.5
[pip3] torch==2.15.0a0+gite67e43b
[pip3] triton==3.7.1
```

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo
